core/src/main/scala/scalacache/package.scala

```diff
@@ -50,21 +59,38 @@ package object scalacache extends JavaSerializationCodec {
           }
       }
 
-      val fromCache: Future[Option[From]] = getWithKey(key).recover[Option[From]] {
-        case e =>
-          if (logger.isWarnEnabled) {
-            logger.warn(s"Failed to read from cache. Key = $key", e)
-          }
-          None
+      def calculateAndCacheResult(): Future[From] = {
+        val result: Future[From] = f
+        asynchronouslyCacheResult(result)
+        result
       }
 
-      fromCache flatMap {
-        case Some(value) => Future.successful(value)
-        case None =>
-          val result: Future[From] = f
-          asynchronouslyCacheResult(result)
-          result
+      val fromCache: Future[Option[From]] = getWithKey(key)
+
+      if (fromCache.isCompleted) {
+        // optimisation for in-memory caches that return Future.successful(...)
+        fromCache.value.get match {
+          case Success(Some(value)) => Future.successful(value)
```

Comment:
I had a quick stab at it but got caught out by the fact that Caffeine does things synchronously whereas we are dealing with Futures in ScalaCache. I would've had to wrap a Future in an `Await.result()`, which is always a bad sign. 

I'd definitely like to revisit this, but it may require a non-trivial redesign of the cache API to more accurately capture the differences between sync and a sync impls. In hindsight, making Caffeine/Guava share exactly the same API as Caffeine/Redis is maybe a bit restrictive.